sklearn/tree/_tree.pyx

```diff
@@ -777,13 +812,37 @@ cdef class Tree:
         cdef Node* node = NULL
         cdef SIZE_t i = 0
 
+        # To handle missing values
+        cdef BOOL_t* missing_mask_ptr = NULL
+        cdef SIZE_t missing_mask_fx_stride
+        cdef UINT32_t* random_state
+        cdef bint allow_missing = <bint>self.allow_missing
+
+        # If allow_missing is True, the missing_mask should be specified.
+        # Either at the higher level
+        if allow_missing:
+            # Extract the missing mask
+            missing_mask_ptr = <BOOL_t*> missing_mask.data
+            missing_mask_fx_stride = <SIZE_t> missing_mask.strides[1] / <SIZE_t> missing_mask.itemsize
+            random_state = &self.rand_r_state
+
         with nogil:
             for i in range(n_samples):
                 node = self.nodes
                 # While node not a leaf
                 while node.left_child != _TREE_LEAF:
                     # ... and node.right_child != _TREE_LEAF:
-                    if X_ptr[X_sample_stride * i +
+                    if allow_missing and missing_mask_ptr[i + missing_mask_fx_stride * node.feature] == 1:
+                        if node.missing_direction == MISSING_DIR_UNDEF:
+                            # Send to either right or left randomly
+                            node = &self.nodes[node.left_child
+                                               if rand_int(0, 2, random_state)
```

Comment:
Okay i now understand why. Yet, we should  pass the random_state object rather than storing it with the tree. This otherwise would lead to some nasty side effects, like:

```
tree = DecisionTreeClassifier(random_state=123)
tree.fit( ... )
tree.predict( ... )  # depends on rng=123
tree.set_params(random_state=456)
tree.predict( ... )  # still depends on rng=123 because we save that reference into the tree structure
```